### PR TITLE
CORTX-28534: Granular deployment (Data-pod-only)

### PR DESCRIPTION
**Pull request by keithpine**
Setting the new `solution.deployment_type` field to `data-only` results in a deployment that only creates CORTX-Data Deployments and skips all of the others: Server, Control and HA. Valid types are `standard` and `data-only`.

The configuration files (`config.yaml` and `cluster.yaml`) are pared down to include only the necessary values to configure the Data and Client pods.

External services (IO and Control) are not created.

Scripts `deploy-cortx-cloud.sh`, `start-cortx-cloud.sh`, and `status-cortx-cloud.sh` are deployment-type-dependent and operate based on the solution YAML file. `destroy-cortx-cloud.sh` and `shutdown-cortx-cloud.sh` are deployment-type-agnostic and gracefully handle any missing resources.

Signed-off-by: Keith Pine <keith.pine@seagate.com>

### CODE REVIEW

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/templates/_cluster.tpl [hunk 1]
```diff
@@ -27,27 +27,34 @@ cluster:
...
     - name: ha
+  {{- end }}
   - name: client_node
     components:
-    - name: hare
+    - name: utils
     - name: motr
       services:
         - motr_client
+    - name: hare
```
Comment (keithpine) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/templates/_cluster.tpl:57]: I updated this to match the latest cluster.yaml.sample file, in an attempt to try and use a client, but I don't think it had any effect.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/templates/_config.tpl [hunk 2]
```diff
@@ -118,12 +122,14 @@ cortx:
...
     clients:
+    {{- if .Values.cortxRgw.enabled }}
       - name: rgw
         num_instances: 1  # number of instances *per-pod*
         endpoints: {{- toYaml .Values.cortxMotr.rgwEndpoints | nindent 8 }}
+    {{- end }}
     {{- if gt (len .Values.cortxMotr.clientEndpoints) 0 }}
       - name: motr_client
-        num_instances: {{ len .Values.cortxMotr.clientEndpoints }}
+        num_instances: {{ .Values.cortxMotr.clientInstanceCount | int }}
```
Comment (keithpine) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/templates/_config.tpl:132]: I believe this is supposed to be the number of instances _per-pod_, not the number of total endpoints. The endpoints value reflects the number of nodes instead.

File: k8_cortx_cloud/destroy-cortx-cloud.sh [hunk 3]
```diff
@@ -229,7 +234,7 @@ function deleteCortxPVs()
                 kubectl delete pv ${pvc_line[0]}
             fi
         fi
-    done <<< "$(kubectl get pv -A)"
```
Comment (keithpine) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:232]: Use `<<<` with a string will cause the while loop to be entered even if the string is empty. This causes errors or misleading messages if the inner loop is not testing for empty strings. Using command substitution `< <()` instead does not have this problem.

File: k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/values.yaml [hunk 4]
```diff
@@ -104,19 +109,21 @@ cortxRgw:
...
 
 #cortxHare allows configuring CORTX HARE settings
 cortxHare:
   haxDataEndpoints: []
   haxServerEndpoints: []
   haxClientEndpoints: []
+  haxService:
+    protocol: https
+    name: cortx-hax-svc
+    port: 22003
```
Comment (keithpine) [inline k8_cortx_cloud/cortx-cloud-helm-pkg/cortx-configmap/values.yaml:122]: HAX is part of Hare, right? That's how it looks to me, not sure why I originally added it to the HA section. Correct me if I'm wrong.

**Comment by keithpine** [comment]
Note that I wasn't able to confirm if clients are working, I just couldn't get anything working there. If there are instructions no how to test this, I'll gladly try.

File: k8_cortx_cloud/destroy-cortx-cloud.sh [hunk 5]
```diff
@@ -297,23 +296,13 @@ function deleteOpenLdap()
...
-        done
+    kubectl get pods --namespace=default --output=name | grep '^pod/openldap-' |
+    while IFS= read -r pod; do
+        kubectl exec "${pod}" --namespace=default -- \
+            bash -c 'rm -rf /etc/3rd-party/* /var/data/3rd-party/* /var/log/3rd-party/*'
+    done
 
-        helm uninstall "openldap" -n "default"
-    fi
+    uninstallHelmChart openldap default
```
Comment (walterlopatka) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:305]: FYI, I think this section (or perhaps using `uninstallHelmchart`) resolves https://jts.seagate.com/browse/CORTX-29468, which I submitted earlier this week.

**Review by walterlopatka — APPROVED** [review]
lgtm.
I salute you, Bash Ninja.

File: k8_cortx_cloud/destroy-cortx-cloud.sh (hunk 5 above)
Comment (keithpine) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:305]: I assigned that to myself and will close when this resolves.

File: k8_cortx_cloud/destroy-cortx-cloud.sh [hunk 6]
```diff
@@ -132,90 +134,93 @@ done <<< "$(kubectl get namespaces)"
 
 num_motr_client=$(extractBlock 'solution.common.motr.num_client_inst')
 
+function uninstallHelmChart()
+{
+    local chart=$1
+    local ns=()
+    [[ -n ${2:-} ]] && ns+=("--namespace=$2")
+    # Silence "release: not found" messages as those can be expected
+    helm uninstall "${chart}" "${ns[@]}" 2> >(grep -v 'uninstall.*release: not found')
```
Comment (osowski) [inline k8_cortx_cloud/destroy-cortx-cloud.sh:143]: Doing some test runs on deploy/destroy loops and catching errors on this command when run it attempts to uninstall the cortx-platform chart.

```bash
########################################################
# Delete Persistent Volumes                            #
########################################################
pvc-1a3c8c01-0e34-4338-9517-12df189811bf
pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089
pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c
Removing pvc-1a3c8c01-0e34-4338-9517-12df189811bf
persistentvolume "pvc-1a3c8c01-0e34-4338-9517-12df189811bf" deleted
Removing pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089
persistentvolume "pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089" deleted
Removing pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c
persistentvolume "pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c" deleted
########################################################
# Delete Cortx Kubernetes Prereqs                      #
########################################################
./destroy-cortx-cloud.sh: line 143: ns[@]: unbound variable
```

Comment (osowski) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:143]: The deploy script exits immediately from this error and leaves the deployment in a somewhat broken state, so I needed to manually clean a bunch of things up both on k8s and in the local fs before I could deploy again.

Comment (keithpine) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:143]: I can't reproduce this error. Could you run the scripts with `bash -x` and send/attach the output?

Comment (osowski) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:143]: The most recent run shows the following:
```
Removing pvc-fa8f7c99-c823-431c-a635-b988d1b32d6f
+ kubectl delete pv pvc-fa8f7c99-c823-431c-a635-b988d1b32d6f
persistentvolume "pvc-fa8f7c99-c823-431c-a635-b988d1b32d6f" deleted
+ [[ default != \d\e\f\a\u\l\t ]]
+ deleteKubernetesPrereqs
+ printf '########################################################\n'
+ printf '# Delete Cortx Kubernetes Prereqs                      #\n'
########################################################
+ printf '########################################################\n'
+ uninstallHelmChart cortx-platform
+ local chart=cortx-platform
# Delete Cortx Kubernetes Prereqs                      #
########################################################
+ ns=()
+ local ns
+ [[ -n '' ]]
./destroy-cortx-cloud.sh: line 143: ns[@]: unbound variable
[root@ssc-vm-g2-rhev4-1987 k8_cortx_cloud]# pwd
/root/deploy-scripts/k8_cortx_cloud
```

Comment (keithpine) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:143]: Ok, I'll revert the `set -u`. This appears to be a problem in bash that was fixed in Bash 4.4+ or so. Locally I'm using Bash 5, the lab servers are using 4.2. 😢

**Comment by keithpine** [comment]
I'm going to update this to also skip creating the IO and Control (CSM) services for a data-only deployment, and also omit the new messages about these services when deployment is complete.

File: k8_cortx_cloud/destroy-cortx-cloud.sh (hunk 6 above)
Comment (osowski) [reply k8_cortx_cloud/destroy-cortx-cloud.sh:141]: > Doing some test runs on deploy/destroy loops and catching errors on this command when run it attempts to uninstall the cortx-platform chart.
> 
> ```shell
> ########################################################
> # Delete Persistent Volumes                            #
> ########################################################
> pvc-1a3c8c01-0e34-4338-9517-12df189811bf
> pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089
> pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c
> Removing pvc-1a3c8c01-0e34-4338-9517-12df189811bf
> persistentvolume "pvc-1a3c8c01-0e34-4338-9517-12df189811bf" deleted
> Removing pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089
> persistentvolume "pvc-7acbe83e-3226-4cb0-8b95-71eccbb5a089" deleted
> Removing pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c
> persistentvolume "pvc-946f9683-90bb-4e62-b5c0-fb9e3599853c" deleted
> ########################################################
> # Delete Cortx Kubernetes Prereqs                      #
> ########################################################
> ./destroy-cortx-cloud.sh: line 143: ns[@]: unbound variable
> ```

As of b66092e, I am no longer seeing this issue. All good on my end.

**Pull request merged**